### PR TITLE
[MM-819]: Added testcases for multiple functions in command_hooks.go file

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**
### Summary
Added testcases for multiple functions in command_hooks.go file

### CODE REVIEW

File: server/command_hooks_test.go
Comment (raghavaggarwal2308) [inline server/command_hooks_test.go:1]: There is a lot of code repetition in this file. Use a loop to minimize that. Like we do in other plugins

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

File: server/command_hooks_test.go [hunk 1]
```diff
@@ -160,6 +160,42 @@ func TestExecuteCommand(t *testing.T) {
...
+		mockLog := mock_pluginapi.NewMockLog(mockCtrl)
+		mockPost := mock_pluginapi.NewMockPost(mockCtrl)
+		mockSlashCommand := mock_pluginapi.NewMockSlashCommand(mockCtrl)
+		mockUser := mock_pluginapi.NewMockUser(mockCtrl)
+		mockSystem := mock_pluginapi.NewMockSystem(mockCtrl)
+		mockConfiguration := mock_pluginapi.NewMockConfiguration(mockCtrl)
+		mockCluster := mock_pluginapi.NewMockCluster(mockCtrl)
+		mockCluster.EXPECT().NewMutex(gomock.Eq(KeyClusterMutex)).Return(pluginapi.NewClusterMutexMock(), nil)
+
+		mockAPI := pluginapi.CustomWrapper(mockChannel, mockFile, mockLog, mockPost, mockSlashCommand, mockUser, mockSystem, mockConfiguration, mockCluster)
```
Comment (raghavaggarwal2308) [inline server/command_hooks_test.go]: Can create a common function for this

File: server/command_hooks_test.go [hunk 2]
```diff
@@ -424,3 +460,145 @@ func TestExecuteCommand(t *testing.T) {
...
+			mockConfiguration := mock_pluginapi.NewMockConfiguration(mockCtrl)
+			mockCluster := mock_pluginapi.NewMockCluster(mockCtrl)
+
+			mockCluster.EXPECT().NewMutex(gomock.Eq(KeyClusterMutex)).Return(pluginapi.NewClusterMutexMock(), nil)
+
+			testCase.setupMocks(mockUser, mockSystem, mockConfiguration)
+
+			mockAPI := pluginapi.CustomWrapper(mockChannel, mockFile, mockLog, mockPost, mockSlashCommand, mockUser, mockSystem, mockConfiguration, mockCluster)
+
+			plugin, _ := setupPlugin(t, mockAPI, time.Now())
```
Comment (raghavaggarwal2308) [inline server/command_hooks_test.go:495]: run api.assertExpectations

Comment (raghavaggarwal2308) [reply server/command_hooks_test.go:495]: Also, what is the second return param here

File: server/command_hooks_test.go [hunk 3]
```diff
@@ -424,3 +460,145 @@ func TestExecuteCommand(t *testing.T) {
...
+		setupMocks             func(mockUser *mock_pluginapi.MockUser, mockSystem *mock_pluginapi.MockSystem, mockConfiguration *mock_pluginapi.MockConfiguration)
+		expectedExportPermission bool
+	}{
+		{
+			name: "user does not have permission to manage channel and is not a system admin",
+			setupMocks: func(mockUser *mock_pluginapi.MockUser, mockSystem *mock_pluginapi.MockSystem, mockConfiguration *mock_pluginapi.MockConfiguration) {
+				mockUser.EXPECT().HasPermissionTo("userID", model.PermissionManageSystem).Return(false)
+				mockUser.EXPECT().HasPermissionToChannel("userID", "chanelID", model.PermissionManageChannelRoles).Return(false)
+			},
+			expectedExportPermission: false,
```
Comment (raghavaggarwal2308) [inline server/command_hooks_test.go]: can remove, check everywhere

File: server/command_hooks_test.go [hunk 4]
```diff
@@ -424,3 +460,145 @@ func TestExecuteCommand(t *testing.T) {
...
+	testCases := []struct {
+		name                   string
+		setupMocks             func(mockUser *mock_pluginapi.MockUser, mockSystem *mock_pluginapi.MockSystem, mockConfiguration *mock_pluginapi.MockConfiguration)
+		expectedExportPermission bool
+	}{
+		{
+			name: "user does not have permission to manage channel and is not a system admin",
+			setupMocks: func(mockUser *mock_pluginapi.MockUser, mockSystem *mock_pluginapi.MockSystem, mockConfiguration *mock_pluginapi.MockConfiguration) {
+				mockUser.EXPECT().HasPermissionTo("userID", model.PermissionManageSystem).Return(false)
+				mockUser.EXPECT().HasPermissionToChannel("userID", "chanelID", model.PermissionManageChannelRoles).Return(false)
```
Comment (raghavaggarwal2308) [inline server/command_hooks_test.go]:
```suggestion
				mockUser.EXPECT().HasPermissionToChannel("mockUserID", "mockChanelID", model.PermissionManageChannelRoles).Return(false)
```
Update everywhere

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

File: server/command_hooks_test.go (hunk 2 above)
Comment (Kshitij-Katiyar) [reply server/command_hooks_test.go:495]: `func setupPlugin(t *testing.T, mockAPI *pluginapi.Wrapper, now time.Time) (*Plugin, *plugin.Context)`

It is plugin context but it is not used in our testcases

File: server/command_hooks_test.go [hunk 5]
```diff
@@ -424,3 +460,144 @@ func TestExecuteCommand(t *testing.T) {
 		wg.Wait()
 	})
 }
+
+func TestHasPermissionToExportChannel(t *testing.T) {
+	testCases := []struct {
+		name                     string
+		setupMocks               func(mockUser *mock_pluginapi.MockUser, mockSystem *mock_pluginapi.MockSystem, mockConfiguration *mock_pluginapi.MockConfiguration)
```
Comment (raghavaggarwal2308) [inline server/command_hooks_test.go]: I don't see the mockSystem and mockConfiguration param being used anywhere

File: server/command_hooks_test.go [hunk 6]
```diff
@@ -160,6 +160,42 @@ func TestExecuteCommand(t *testing.T) {
...
+		mockChannel := mock_pluginapi.NewMockChannel(mockCtrl)
+		mockFile := mock_pluginapi.NewMockFile(mockCtrl)
+		mockLog := mock_pluginapi.NewMockLog(mockCtrl)
+		mockPost := mock_pluginapi.NewMockPost(mockCtrl)
+		mockSlashCommand := mock_pluginapi.NewMockSlashCommand(mockCtrl)
+		mockUser := mock_pluginapi.NewMockUser(mockCtrl)
+		mockSystem := mock_pluginapi.NewMockSystem(mockCtrl)
+		mockConfiguration := mock_pluginapi.NewMockConfiguration(mockCtrl)
+		mockCluster := mock_pluginapi.NewMockCluster(mockCtrl)
+		mockCluster.EXPECT().NewMutex(gomock.Eq(KeyClusterMutex)).Return(pluginapi.NewClusterMutexMock(), nil)
```
Comment (raghavaggarwal2308) [inline server/command_hooks_test.go]: can use setup func

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

**Review by raghavaggarwal2308 — APPROVED** [review]

**Comment by Kshitij-Katiyar** [comment]
Created a PR with all these changes on MM remote hence closing this one

**Pull request closed without merging**